File: challenges/week1.longest-substring-with-repetition/ikasty.js
/**
 * Runtime: 168 ms, faster than 25.32% of JavaScript online submissions.
 * Memory Usage: 46.9 MB, less than 15.02% of JavaScript online submissions.
 */

/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var longestSubstring = function(s, k) {
    let count = new Array(s.length)
    count[-1] = {} // 🤣🔫

    let idxmap = {} // 마지막으로 글자가 등장한 idx
    for (let i = 0; i < s.length; i++) {
        let target = s[i]
        count[i] = {}
        if (target in idxmap)
            count[i][target] = count[idxmap[target]][target] + 1
        else
            count[i][target] = 1
        idxmap[target] = i
    }

    let checkDiff = function(target, st, ed) {
        let lastIdx = (idx) => {
            // 여기에서는 idxmap을 사용할 수 없음
            for (; idx >= 0 && !(target in count[idx]); idx--);
            return idx
        }

        // memoize
        count[ed][target] = count[lastIdx(ed)][target] || 0
        count[st][target] = count[lastIdx(st)][target] || 0
        return count[ed][target] - count[st][target]
    }

    let max = 0
    for (let start = 0; start < s.length - max; start++) {
        let checklist = new Set(s.substr(start, max).split(''))
        let donelist  = new Set()

        endloop:
        for (let end = start + max; end < s.length; end++) {
            checklist.add(s[end])

            for (let c of checklist) {
                // 더 이상 c를 검사할 필요가 없음
                if (donelist.has(c)) continue

                // c에서 실패함
                if (checkDiff(c, start-1, end) < k) continue endloop

                // c에서 성공
                donelist.add(c)
            }

            if (end - start + 1 > max) max = end - start + 1
        }

        // 다른 글자가 나올 때까지 스킵한다
        while (start < s.length - k && s[start+1] == s[start]) start++
    }

    return max
};

File: challenges/week6.swim-in-rising-water/ikasty.py
# Runtime: 88 ms, faster than 98.70% of Python3 online submissions
# Memory Usage: 14.5 MB, less than 98.27% of Python3 online submissions

class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        N = len(grid)

        def check_hole(step):
            # 정상 "웅덩이" 인지 체크
            def valid_hole(x, y):
                if x < 0 or x >= N:
                    return False
                if y < 0 or y >= N:
                    return False
                if grid[x][y] > step:
                    return False
                return True

            stack = [(0, 0)]
            hole = set([(0, 0)])
            while len(stack) > 0:
                i, j = stack.pop()
                if (i, j) == (N-1, N-1):
                    return True

                for x, y in [(i-1, j), (i, j-1), (i+1, j), (i, j+1)]:
                    if valid_hole(x, y) and (x, y) not in hole:
                        hole.add((x, y))
                        stack.append((x, y))

            return False

        # start == 실패한 step
        start = max(N*2-2, grid[0][0], grid[-1][-1])
        # end   == 성공한 step
        end = N*N - 1

        # init check
        if check_hole(start):
            return start

        while (start + 1 < end):
            step = start + int((end - start) / 2)
            if check_hole(step):
                end = step
            else:
                start = step

        return end
